File: src/test/test_angle.f90
program test_angle
  use control
  use angle

  implicit none

  integer :: l = 7, t1=1, t2=1, t3=1, t4=1, testCond
  double precision :: mu_test(7), wt_test(7)

  mu_test = [0.9862838086968123,0.9284348836635735,0.8272013150697650,0.6872929048116855,0.5152486363581541,0.3191123689278897,0.1080549487073437]
  wt_test = [0.0351194603317519,0.0801580871597602,0.1215185706879032,0.1572031671581935,0.1855383974779378,0.2051984637212956,0.2152638534631578]

  call initialize_control('test/reg_test_options', .true.)
  angle_order = 7
  angle_option = 1
  call initialize_angle()
  t1 = testCond(norm2(mu-mu_test) < 1e-12)
  t2 = testCond(norm2(wt-wt_test) < 1e-12)
  call finalize_angle()
  call finalize_control()

  mu_test = [0.974553956171379190, 0.87076559279969723, 0.70292257568869854, 0.50000000000000000, 0.29707742431130146, 0.12923440720030277, 0.025446043828620812]
  wt_test = [0.064742483084434838, 0.13985269574463835, 0.19091502525255952, 0.20897959183673470, 0.19091502525255952, 0.13985269574463835, 0.064742483084434838]
  
  call initialize_control('test/reg_test_options', .true.)
  angle_order = 7
  angle_option = 2
  call initialize_angle()
  t3 = testCond(norm2(mu-mu_test) < 1e-12)
  t4 = testCond(norm2(wt-wt_test) < 1e-12)
  call finalize_angle()
  call finalize_control()
  
  if (t1 == 0) then
    print *, 'angle: GL mu failed'
  else if (t2 == 0) then
    print *, 'angle: GL wt failed'
  else if (t3 == 0) then
    print *, 'angle: DGL mu failed'
  else if (t4 == 0) then
    print *, 'angle: DGL wt failed'
  else
    print *, 'all tests passed for angle'
  end if

end program test_angle

integer function testCond(condition)
  logical, intent(in) :: condition
  if (condition) then
    write(*,"(A)",advance="no") '.'
    testCond = 1
  else
    write(*,"(A)",advance="no") 'F'
    testCond = 0
  end if

end function testCond

File: src/test/test_angle.py
import sys
sys.path.append('../')

import unittest
import pydgm
import numpy as np
from scipy.special import sph_harm


class TestANGLE_1D(unittest.TestCase):

    def setUp(self):
        pydgm.control.spatial_dimension = 1

    def test_angle_legendre_p(self):
        ''' 
        Test the function used to create the legendre polynomials
        '''

        angle = pydgm.angle

        self.assertEqual(angle.legendre_p(0, 0.5), 1)
        self.assertEqual(angle.legendre_p(1, 0.5), 0.5)
        self.assertEqual(angle.legendre_p(2, 0.5), -0.125)

    def test_angle_d_legendre_p(self):
        ''' 
        Test the function used to create the double legendre polynomials
        '''

        angle = pydgm.angle

        self.assertEqual(angle.d_legendre_p(0, 0.5), 0.0)
        self.assertEqual(angle.d_legendre_p(1, 0.5), 1.0)
        self.assertEqual(angle.d_legendre_p(2, 0.5), 1.5)

    def test_angle_initialize_polynomials(self):
        '''
        Test the the legendre polynomial basis is correct
        '''
        nAngle = 8
        pydgm.control.angle_order = nAngle
        pydgm.control.angle_option = pydgm.angle.gl
        pydgm.control.number_legendre = 7

        pydgm.angle.initialize_angle()
        pydgm.angle.initialize_polynomials()

        # Get the basis set
        basis = pydgm.angle.p_leg

        # Use numpy to get the test basis
        x, wt = np.polynomial.legendre.leggauss(nAngle * 2)
        I = np.eye(nAngle * 2)
        test_basis = np.array([np.polynomial.legendre.legval(x, I[i]) * (-1) ** i for i in range(nAngle)])

        # Test the basis
        np.testing.assert_array_almost_equal(basis, test_basis, 12)

    def test_angle_quadrature(self):
        '''
        Test for correct gauss legendre points and weights
        '''
        nAngle = 8
        # setup control parameters for quadrature
        pydgm.control.angle_order = nAngle
        pydgm.control.angle_option = pydgm.angle.gl

        # build quadrature, test it, and deallocate
        pydgm.angle.initialize_angle()

        # Get the test quadrature weights from numpy
        mu_test, wt_test = np.polynomial.legendre.leggauss(nAngle * 2)
        mu_test = mu_test[:nAngle - 1:-1]
        wt_test = wt_test[:nAngle - 1:-1]

        # Test for equivalance
        np.testing.assert_array_almost_equal(pydgm.angle.mu, mu_test, 12)
        np.testing.assert_array_almost_equal(pydgm.angle.wt, wt_test, 12)

    def tearDown(self):
        pydgm.angle.finalize_angle()
        pydgm.control.finalize_control()


class TestANGLE_2D(unittest.TestCase):

    def setUp(self):
        pydgm.control.spatial_dimension = 2
        pydgm.control.angle_order = 10

    def test_angle_quadrature(self):
        '''
        Test for correct gauss legendre points and weights
        '''
        # setup control parameters for quadrature
        pydgm.control.angle_option = pydgm.angle.gl

        # build quadrature, test it, and deallocate
        pydgm.angle.initialize_angle()

        mu = [0.189321326478, 0.189321326478, 0.189321326478, 0.189321326478, 0.189321326478, 0.508881755583, 0.508881755583, 0.508881755583, 0.508881755583, 0.694318887594, 0.694318887594, 0.694318887594, 0.839759962237, 0.839759962237, 0.96349098111]
        eta = [0.189321326478, 0.508881755583, 0.694318887594, 0.839759962237, 0.96349098111 , 0.189321326478, 0.508881755583, 0.694318887594, 0.839759962237, 0.189321326478, 0.508881755583, 0.694318887594, 0.189321326478, 0.508881755583, 0.189321326478]
        wt = [0.089303147984, 0.072529151712, 0.045043767436, 0.072529151712, 0.089303147984, 0.072529151712, 0.053928114488, 0.053928114488, 0.072529151712, 0.045043767436, 0.053928114488, 0.045043767436, 0.072529151712, 0.072529151712, 0.089303147984]

        # Test for equivalance
        np.testing.assert_array_almost_equal(pydgm.angle.mu, mu, 12)
        np.testing.assert_array_almost_equal(pydgm.angle.eta, eta, 12)
        np.testing.assert_array_almost_equal(pydgm.angle.wt / 0.5 / np.pi, wt, 12)

    def tearDown(self):
        pydgm.angle.finalize_angle()
        pydgm.control.finalize_control()


if __name__ == '__main__':

    unittest.main()

